src/wp-includes/html-api/class-wp-html-processor.php

```diff
@@ -0,0 +1,259 @@
+<?php
+
+class WP_HTML_Processor extends WP_HTML_Tag_Processor {
+	public $fully_supported_input = null;
+	public $open_elements = array();
+
+	public function ensure_support() {
+		if ( null !== $this->fully_supported_input ) {
+			return $this->fully_supported_input;
+		}
+
+		$stack = array();
+
+		$p = new WP_HTML_Tag_Processor( $this->html );
+		while ( $p->next_tag( array( 'tag_closers' => 'visit' ) ) ) {
+			$tag_name = $p->get_tag();
+
+			if ( ! $p->is_tag_closer() ) {
+				$element = WP_HTML_Spec::element_info( $tag_name );
+
+				$self_closes = $element::is_void || ( ! $element::is_html && $p->has_self_closing_flag() );
+				if ( ! $self_closes ) {
+					$stack[] = $tag_name;
+				}
+			} else {
+				if ( end( $stack ) === $tag_name ) {
+					array_pop( $stack );
+					continue;
+				}
+
+				$this->fully_supported_input = false;
+				return false;
+			}
+		}
+
+		$this->fully_supported_input = 0 === count( $stack );
+
+		return $this->fully_supported_input;
+	}
+
+	public function next_tag( $query = null ) {
```

Comment:
I wish we could make it private in the descendant class. It doesn't even make sense to have it public, it can't support the same set of queries as the parent, and it's a pain to deal with PHP's idea of polymorphism where the parent class will call the child implementation of this method anyway.

Reply:
funny enough I added `_doing_it_wrong()` here before realizing that I call it from within this tag processor.

is it the tag processor that's calling this?

the one thing I wish we could do is remove `$query` but it wouldn't let me. still some thinking about this one to do.